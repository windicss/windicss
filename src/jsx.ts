/* eslint-disable @typescript-eslint/no-empty-interface */
export type UtilityNames =
  | 'p'
  | 'm'
  | 'w'
  | 'h'
  | 'z'
  | 'border'
  | 'grid'
  | 'flex'
  | 'bg'
  | 'text'
  | 'font'
  | 'opacity'
  | 'animate'
  | 'transition'
  | 'transform'
  | 'align'
  | 'justify'
  | 'items'
  | 'block'
  | 'content'
  | 'pos'
  | 'box'
  | 'overflow'
  | 'underline'
  | 'list'
  | 'gradient'
  | 'divide'
  | 'gap'
  | 'ring'
  | 'icon'
  | 'container'
  | 'space'
  | 'table'
  | 'order'
  | 'place'
  | 'display'
  | 'shadow'
  | 'blend'
  | 'filter'
  | 'backdrop'
  | 'cursor'
  | 'outline'
  | 'select'

export type VariantNames =
  | 'hover'
  | 'active'
  | 'focus'
  | 'enabled'
  | 'dark'
  | 'light'
  | 'sm'
  | 'lg'
  | 'md'
  | 'xl'
  | 'xxl'
  | 'first'
  | 'last'
  | 'child'
  | 'root'
  | 'before'
  | 'after'
  | 'all'

export type AttributifyNames<Prefix extends string = ''> =
  | `${Prefix}${UtilityNames}`
  | `${Prefix}${VariantNames}`
  | `${Prefix}${VariantNames}:${UtilityNames}`

export interface AttributifyAttributes extends Partial<Record<AttributifyNames, string>> {}
