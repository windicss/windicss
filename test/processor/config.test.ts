import { resolve } from 'path';
import { toType } from '../../src/utils/tools';
import { Processor } from '../../src/lib';
import { twExclude } from '../../src/config';
import { CSSParser } from '../../src/utils/parser';
import type { colorObject } from '../../src/interfaces';

const configPath = resolve('./test/assets/windi.config.js');
const userConfig = require(configPath);

describe('Config', () => {
  const baseConfig = new Processor();

  it('dict input', () => {
    const processor = new Processor(userConfig);
    expect(processor.config('theme.screens')).toEqual(
      processor.theme('screens')
    );
    expect(processor.theme('screens')).toEqual(userConfig.theme.screens);
    expect(processor.theme('colors')).toEqual(userConfig.theme.colors);
    expect(processor.theme('colors.pink')).toEqual(
      userConfig.theme.colors.pink
    );
    expect(processor.theme('fontFamily')).toEqual(userConfig.theme.fontFamily);
    expect(processor.theme('spacing')).toEqual({
      ...toType(baseConfig.theme('spacing'), 'object'),
      ...userConfig.theme.extend.spacing,
    });
    expect(processor.theme('borderRadius')).toEqual({
      ...toType(baseConfig.theme('borderRadius'), 'object'),
      ...userConfig.theme.extend.borderRadius,
    });
  });

  it('resolveConfig should extend correctly', () => {
    const processor = new Processor({
      theme: {
        extend: {
          order: {
            lg: '44',
          },
          lineClamp: {
            sm: '4',
            lg: '9',
          },
        },
      },
    });
    expect(processor.theme('order')).toMatchSnapshot('order');
    expect(processor.theme('lineClamp')).toEqual({ sm: '4', lg: '9' });
  });

  it('resolveConfig should not overwrite theme function', () => {
    const processor = new Processor({
      theme: {
        extend: {
          colors: {
            red: {
              custom: '#ea1851',
            },
          },
          width: {
            '1/7': '14%',
            '2/7': '28%',
            '3/7': '42%',
            '4/7': '57%',
            '5/7': '71%',
            '6/7': '85%',
          },
        },
      },
    });
    expect(processor.theme('width')).toMatchSnapshot('extend-width');
    // theme.extend.colors overwrites defaults #141
    expect(processor.theme('colors.red')).toMatchSnapshot('extend-colors-red');
  });

  it('user theme should overwrite default theme', () => {
    const processor = new Processor({
      theme: {
        colors: {
          white: '#ffffff',
        },
      },
    });
    expect(processor.theme('colors')).toEqual({
      white: '#ffffff',
    });
  });

  it('change separator test', () => {
    const processor = new Processor({ separator: '_' });
    expect(processor.interpret('sm_bg-gray-200').styleSheet.build()).toMatchSnapshot('_');
    processor.loadConfig({ separator: '::' });
    expect(processor.interpret('sm::bg-gray-200').styleSheet.build()).toMatchSnapshot('::');
    processor.loadConfig({ separator: '-' });
    expect(processor.interpret('sm-bg-gray-200').styleSheet.build()).toMatchSnapshot('-');
    processor.loadConfig({ separator: '--' });
    expect(processor.interpret('sm--bg-gray-200').styleSheet.build()).toMatchSnapshot('--');
    processor.loadConfig({ separator: '---' });
    expect(processor.interpret('sm---bg-gray-200').styleSheet.build()).toMatchSnapshot('---');
    processor.loadConfig({ separator: '_' });
    expect(processor.interpret('sm_bg-gray-200').styleSheet.build()).toMatchSnapshot('_');
    processor.loadConfig({ separator: '__' });
    expect(processor.interpret('sm__bg-gray-200').styleSheet.build()).toMatchSnapshot('__');
    processor.loadConfig({ separator: '.' });
    expect(processor.interpret('sm.bg-gray-200').styleSheet.build()).toMatchSnapshot('.');
  });

  it('add prefix test', () => {
    const processor = new Processor({ prefix: 'tw-' });
    expect(processor.interpret('sm:tw-bg-black').styleSheet.build()).toBe(
      '@media (min-width: 640px) {\n  .sm\\:tw-bg-black {\n    --tw-bg-opacity: 1;\n    background-color: rgba(0, 0, 0, var(--tw-bg-opacity));\n  }\n}'
    );
  });

  it('does not generate non-prefixed classes when using prefix', () => {
    const processor = new Processor({ prefix: 'windi-' });
    const classes = 'items-center justify-center flex-wrap block flex windi-block';
    // it should not compile the standard classes (items-center, justify-center) because they are not prefixed
    expect(processor.interpret(classes).styleSheet.build()).toBe(
      '.windi-block {\n  display: block;\n}'
    );
  });


  it('important test', () => {
    const processor = new Processor({ important: true });
    expect(processor.interpret('sm:bg-black').styleSheet.build()).toBe(
      '@media (min-width: 640px) {\n  .sm\\:bg-black {\n    --tw-bg-opacity: 1 !important;\n    background-color: rgba(0, 0, 0, var(--tw-bg-opacity)) !important;\n  }\n}'
    );
  });

  it('important string test', () => {
    const processor = new Processor({ important: '#app' });
    expect(processor.interpret('sm:bg-black').styleSheet.build()).toMatchSnapshot('css');
    expect(processor.interpret('dark:bg-white').styleSheet.build()).toMatchSnapshot('css');
  });

  it('animation config test', () => {
    const processor = new Processor({
      theme: {
        extend: {
          animation: {
            skbounce: 'skbounce 2.0s infinite ease-in-out',
          },
          keyframes: {
            skbounce: {
              '0%, 100%': { transform: 'scale(0.0)' },
              '50%': { transform: 'scale(1.0)' },
            },
          },
        },
      },
    });
    expect(processor.interpret('animate-skbounce').styleSheet.build()).toMatchSnapshot('animate');
  });

  it('color config test', () => {
    const processor = new Processor({
      darkMode: 'media', // or 'media' or 'class'
      theme: {
        extend: {
          colors: {
            active: 'var(--active)',
            primary: 'var(--primary)',
            'on-primary': 'var(--on-primary)',
            'on-primary-active': 'var(--on-primary-active)',
            frame: 'var(--frame)',
            'on-frame': 'var(--on-frame)',
            'on-frame-active': 'var(--on-frame-active)',
            darkTheme: {
              600: '#262A34',
              700: '#181A20',
              800: '#1A1B20',
            },
          },
        },
      },
    });
    expect(processor.theme('colors.darkTheme')).toEqual({
      600: '#262A34',
      700: '#181A20',
      800: '#1A1B20',
    });

    expect(processor.theme('colors.frame')).toEqual('var(--frame)');

    const styleSheet = processor.interpret('bg-darkTheme-600 bg-active bg-on-primary bg-on-frame-active').styleSheet;
    expect(styleSheet.build()).toMatchSnapshot('css');
  });

  it('handle colors test', () => {
    const processor = new Processor({
      theme: {
        extend: {
          colors: {
            discord: {
              DEFAULT: '#7289da',
              '100': '#7289da',
            },
            'primary-color-red': '#FF0000',
          },
        },
      },
    });
    const styleSheet = processor.interpret(
      'bg-discord bg-discord-100 bg-hex-7289da ring-offset-hex-1c1c1e ring-offset-gray-200 text-primary-color-red'
    ).styleSheet;
    expect(styleSheet.build()).toMatchSnapshot('css');
  });

  it('handle nest colors', () => {
    const processor = new Processor({
      theme: {
        extend: {
          colors: {
            my: {
              DEFAULT: '#000',
              custom: {
                red: '#ff0000',
                DEFAULT: '#fff',
              },
            },
          },
        },
      },
    });

    const styleSheet = processor.interpret('bg-my-custom-red bg-my bg-my-custom text-my-custom').styleSheet;
    expect(styleSheet.build()).toMatchSnapshot('colors');
  });

  it('exclude config', () => {
    const processor = new Processor({
      theme: {
        extend: {
          colors: {
            discord: {
              DEFAULT: '#7289da',
              '100': '#7289da',
            },
          },
        },
      },
      exclude: twExclude,
    });
    expect(processor.interpret('bg-hex-1c1c1e bg-$test-variable p-1rem p-4px p-3 p-4.2 sm:p-4.5').ignored.length).toEqual(7);
  });

  it('shortcuts config string', () => {
    const processor = new Processor({
      shortcuts: {
        'btn': 'py-2 px-4 font-semibold rounded-lg shadow-md',
        'btn-green': 'text-white bg-green-500 hover:bg-green-700',
      },
    });
    expect(processor.interpret('btn btn-green lg:btn').styleSheet.build()).toMatchSnapshot('shortcuts string');
  });

  it('shortcuts config object', () => {
    const processor = new Processor({
      shortcuts: {
        'btn': {
          'color': 'white',
          '@apply': 'py-2 px-4 font-semibold rounded-lg',
          '&:hover': {
            '@apply': 'bg-green-700',
            'color': 'black',
          },
        },
      },
    });
    expect(processor.interpret('btn').styleSheet.build()).toMatchSnapshot('shortcuts object');
  });

  it('allows to use prefix with shortcuts', () => {
    const processor = new Processor({
      prefix: 'windi-',
      shortcuts: {
        'btn': 'py-2 px-4 font-semibold rounded-lg shadow-md',
        'btn-green': {
          '@apply': 'text-white bg-green-500 hover:bg-green-700',
        },
      },
    });
    const result = processor.interpret('windi-btn windi-btn-green');
    expect(result.ignored.length).toEqual(0);
    expect(result.styleSheet.build()).toMatchSnapshot('shortcuts with prefix');
  });

  it('different colors', () => {
    const processor = new Processor({
      theme: {
        colors: {
          name: 'blue',
          hex: '#1c1c1e',
          hexa: '#0000ff00',
          rgb: 'rgb(23, 23, 24)',
          rgba: 'rgba(23, 23, 25, 0.5)',
          hsl: 'hsl(120, 100%, 50%)',
          hsla: 'hsla(120, 100%, 75%, 0.3)',
          hwb: 'hwb(280, 40%, 60%)',
          hwba: 'hwb(280, 40%, 60%, 0.7)',
          srgb: 'rgb(69.99%, 32%, 32%)',
        },
      },
    });
    expect(processor.interpret('bg-name bg-hex bg-hexa bg-rgb bg-rgba bg-hsl bg-hsla bg-hwb bg-hwba bg-srgb').styleSheet.build()).toMatchSnapshot('colors');
  });

  it('close corePlugins', () => {
    const processor = new Processor({
      corePlugins: {
        container: false,
        animation: false,
      },
    });
    expect(processor.interpret('container sm:container animate-ping bg-red-200').ignored.length).toEqual(3);
    expect(processor.interpret('bg-red-500').success.length).toEqual(1);
  });

  it('set corePlugins', () => {
    const processor = new Processor({
      corePlugins: [
        'cursor',
        'backgroundColor',
      ],
    });
    expect(processor.interpret('cursor-pointer bg-red-200 bg-green-300 font-bold container').success.length).toEqual(3);
  });

  it('variables theme config', () => {
    const processor = new Processor({
      theme: {
        vars: {
          primary: '#6366f1',
        },
      },
    });
    expect(processor.preflight(undefined, false, false, true).build()).toMatchSnapshot('css');
  });

  it('config for alpha channel', () => {
    const processor = new Processor({
      theme: {
        extend: {
          colors: {
            background: 'rgba(255, 0, 0, 0.5)',
          },
        },
      },
    });
    expect(processor.interpret('bg-background').styleSheet.build()).toEqual(`.bg-background {
  --tw-bg-opacity: 0.5;
  background-color: rgba(255, 0, 0, var(--tw-bg-opacity));
}`);
  });

  it('presets test', () => {
    const processor = new Processor({
      presets: [
        {
          theme: {
            colors: {
              blue: {
                light: '#85d7ff',
                DEFAULT: '#1fb6ff',
                dark: '#009eeb',
              },
              pink: {
                light: '#ff7ce5',
                DEFAULT: '#ff49db',
                dark: '#ff16d1',
              },
            },
            extend: {
              flexGrow: {
                2: '20',
              },
            },
          },
        },
        {
          theme: {
            colors: {
              pink: {
                light: '#ff7ce5',
                dark: '#ff16d1',
              },
            },
            extend: {
              flexGrow: {
                3: '30',
              },
            },
          },
        },
      ],
      theme: {
        fontFamily: {
          sans: ['Graphik', 'sans-serif'],
        },
        extend: {
          minHeight: {
            48: '12rem',
          },
        },
      },
    });
    // same key should be overwrite
    expect(processor.theme('colors.blue.light')).toBeUndefined();
    expect(processor.theme('colors.blue.400')).toBeUndefined();
    expect(processor.theme('colors.pink.DEFAULT')).toBeUndefined();
    expect(processor.theme('colors.pink.dark')).toEqual('#ff16d1');
    expect(processor.theme('colors.pink.light')).toEqual('#ff7ce5');
    expect(processor.theme('fontFamily.sans')).toEqual([ 'Graphik', 'sans-serif' ]);

    // default value should be overwrite
    expect(processor.theme('colors.pink.500')).toBeUndefined();
    expect(processor.theme('colors.indigo.500')).toBeUndefined();
    expect(processor.theme('fontFamily.mono')).toBeUndefined();

    // extend key should be combined
    expect(processor.theme('minHeight.prose')).toEqual('65ch');
    expect(processor.theme('minHeight.48')).toEqual('12rem');
    expect(processor.theme('flexGrow.0')).toEqual('0');
    expect(processor.theme('flexGrow.2')).toEqual('20');
    expect(processor.theme('flexGrow.3')).toEqual('30');
  });

  it('extend black', () => {
    const processor = new Processor({
      theme: {
        extend: {
          colors: {
            black : {
              300: '#1c1c1e',
              400: '#200',
            },
          },
        },
      },
    });
    expect(processor.interpret('bg-black-300 bg-black bg-black-400').styleSheet.build()).toMatchSnapshot('css');
  });

  it('config with brackets', () => {
    const processor = new Processor({
      theme: {
        extend: {
          height: theme => ({
            '(custom)': `calc(100% - ${theme('spacing.16')})`,
          }),
        },
      },
    });
    expect(processor.interpret('h-(custom) text-lg font-bold').styleSheet.build()).toMatchSnapshot('css');
  });

  // #218
  it('dashify', () => {
    const processor = new Processor({
      shortcuts: {
        foo: {
          '-TwBgOpacity': '1',
        },
        bar: {
          'MozUserSelect': '1',
        },
        caps: {
          'testCss': '1',
        },
      },
    });
    expect(processor.interpret('foo bar caps').styleSheet.build()).toMatchSnapshot('css');
  });

  // #264
  it('transition property', () => {
    const processor = new Processor({
      theme: {
        extend: {
          transitionProperty: {
            background: 'background-color',
          },
        },
      },
    });
    const parser = new CSSParser(`
    .btn {
      @apply transition-background;
    }
    `, processor);
    expect(parser.parse().build()).toMatchSnapshot('css');
  });

  // #246
  it('confuses color with font-size if keys are the same', () => {
    const processor = new Processor({
      theme: {
        colors: {
          gray: {
            DEFAULT: '#fff',
            base: '#59AE90',
          },
        },
        fontSize: {
          base: '125%',
        },
      },
    });
    expect(processor.interpret('text-gray text-gray-base text-base').styleSheet.build()).toMatchSnapshot('css');
  });

  it('support color callback function #75 #256', () => {
    const colorObject: colorObject = {
      'transparent': 'transparent',
      'current': 'currentColor',
    };
    const colorValues = {
      'primary'           : { name: 'p', value: '259 94% 51%' },
      'primary-focus'     : { name: 'pf', value: '259 94% 41%' },
      'primary-content'   : { name: 'pc', value: '0 0% 100%' },
    };
    for (const [key, { name, value }] of Object.entries(colorValues)) {
      colorObject[key] = ({ opacityVariable, opacityValue }) => {
        if (opacityValue !== undefined) return `hsla(var(--${name}, ${value}) / ${opacityValue})`;
        if (opacityVariable !== undefined) return `hsla(var(--${name}, ${value}) / var(${opacityValue}, 1))`;
        return `hsl(var(--${name}, ${value}))`;
      };
    }
    const processor = new Processor({
      theme: {
        colors: colorObject,
      },
    });
    expect(processor.interpret('bg-transparent bg-primary bg-primary-focus text-primary-content').styleSheet.build()).toMatchSnapshot('css');
  });

  // #402
  it('container with screens', () => {
    const processor = new Processor({
      theme: {
        screens: {
          print: { raw: 'print' },
        },
      },
    });
    expect(processor.interpret('print:bg-black container').styleSheet.build()).toMatchSnapshot('css');
  });
});
