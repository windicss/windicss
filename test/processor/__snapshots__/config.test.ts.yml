Tools / allows to use prefix with shortcuts / shortcuts with prefix / 0: |-
  .windi-btn {
    border-radius: 0.5rem;
    font-weight: 600;
    padding-top: 0.5rem;
    padding-bottom: 0.5rem;
    padding-left: 1rem;
    padding-right: 1rem;
    --tw-shadow: 0 4px 6px -1px rgb(0 0 0/0.1),0 2px 4px -2px rgb(0 0 0/0.1);
    --tw-shadow-colored: 0 4px 6px -1px var(--tw-shadow-color),0 2px 4px -2px var(--tw-shadow-color);
    -webkit-box-shadow: var(--tw-ring-offset-shadow,0 0 #0000),var(--tw-ring-shadow,0 0 #0000),var(--tw-shadow);
    box-shadow: var(--tw-ring-offset-shadow,0 0 #0000),var(--tw-ring-shadow,0 0 #0000),var(--tw-shadow);
  }
  .windi-btn-green {
    --tw-bg-opacity: 1;
    background-color: rgba(16, 185, 129, var(--tw-bg-opacity));
    --tw-text-opacity: 1;
    color: rgba(255, 255, 255, var(--tw-text-opacity));
  }
  .windi-btn-green:hover {
    --tw-bg-opacity: 1;
    background-color: rgba(4, 120, 87, var(--tw-bg-opacity));
  }
Tools / animation config test / animate / 0: |-
  @keyframes skbounce {
    0%, 100% {
      -webkit-transform: scale(0.0);
      transform: scale(0.0);
    }
    50% {
      -webkit-transform: scale(1.0);
      transform: scale(1.0);
    }
  }
  @-webkit-keyframes skbounce {
    0%, 100% {
      -webkit-transform: scale(0.0);
      transform: scale(0.0);
    }
    50% {
      -webkit-transform: scale(1.0);
      transform: scale(1.0);
    }
  }
  .animate-skbounce {
    -webkit-animation: skbounce 2.0s infinite ease-in-out;
    animation: skbounce 2.0s infinite ease-in-out;
  }
Tools / change separator test / - / 2: |-
  @media (min-width: 640px) {
    .sm-bg-gray-200 {
      --tw-bg-opacity: 1;
      background-color: rgba(229, 231, 235, var(--tw-bg-opacity));
    }
  }
Tools / change separator test / -- / 3: |-
  @media (min-width: 640px) {
    .sm--bg-gray-200 {
      --tw-bg-opacity: 1;
      background-color: rgba(229, 231, 235, var(--tw-bg-opacity));
    }
  }
Tools / change separator test / --- / 4: |-
  @media (min-width: 640px) {
    .sm---bg-gray-200 {
      --tw-bg-opacity: 1;
      background-color: rgba(229, 231, 235, var(--tw-bg-opacity));
    }
  }
Tools / change separator test / . / 7: |-
  @media (min-width: 640px) {
    .sm\.bg-gray-200 {
      --tw-bg-opacity: 1;
      background-color: rgba(229, 231, 235, var(--tw-bg-opacity));
    }
  }
'Tools / change separator test / :: / 1': |-
  @media (min-width: 640px) {
    .sm\:\:bg-gray-200 {
      --tw-bg-opacity: 1;
      background-color: rgba(229, 231, 235, var(--tw-bg-opacity));
    }
  }
Tools / change separator test / _ / 0: |-
  @media (min-width: 640px) {
    .sm_bg-gray-200 {
      --tw-bg-opacity: 1;
      background-color: rgba(229, 231, 235, var(--tw-bg-opacity));
    }
  }
Tools / change separator test / _ / 5: |-
  @media (min-width: 640px) {
    .sm_bg-gray-200 {
      --tw-bg-opacity: 1;
      background-color: rgba(229, 231, 235, var(--tw-bg-opacity));
    }
  }
Tools / change separator test / __ / 6: |-
  @media (min-width: 640px) {
    .sm__bg-gray-200 {
      --tw-bg-opacity: 1;
      background-color: rgba(229, 231, 235, var(--tw-bg-opacity));
    }
  }
Tools / color config test / css / 0: |-
  .bg-darkTheme-600 {
    --tw-bg-opacity: 1;
    background-color: rgba(38, 42, 52, var(--tw-bg-opacity));
  }
  .bg-active {
    background-color: var(--active);
  }
  .bg-on-primary {
    background-color: var(--on-primary);
  }
  .bg-on-frame-active {
    background-color: var(--on-frame-active);
  }
Tools / config with brackets / css / 0: |-
  .font-bold {
    font-weight: 700;
  }
  .h-\(custom\) {
    height: calc(100% - 4rem);
  }
  .text-lg {
    font-size: 1.125rem;
    line-height: 1.75rem;
  }
Tools / confuses color with font-size if keys are the same / css / 0: |-
  .text-base {
    font-size: 125%;
  }
  .text-gray {
    --tw-text-opacity: 1;
    color: rgba(255, 255, 255, var(--tw-text-opacity));
  }
  .text-gray-base {
    --tw-text-opacity: 1;
    color: rgba(89, 174, 144, var(--tw-text-opacity));
  }
Tools / container with screens / css / 0: |-
  .container {
    width: 100%;
  }
  @media print {
    .print\:bg-black {
      --tw-bg-opacity: 1;
      background-color: rgba(0, 0, 0, var(--tw-bg-opacity));
    }
  }
Tools / dashify / css / 0: |-
  .foo {
    --tw-bg-opacity: 1;
  }
  .bar {
    -moz-user-select: 1;
  }
  .caps {
    test-css: 1;
  }
Tools / different colors / colors / 0: |-
  .bg-name {
    --tw-bg-opacity: 1;
    background-color: rgba(0, 0, 255, var(--tw-bg-opacity));
  }
  .bg-hex {
    --tw-bg-opacity: 1;
    background-color: rgba(28, 28, 30, var(--tw-bg-opacity));
  }
  .bg-hexa {
    --tw-bg-opacity: 0;
    background-color: rgba(0, 0, 255, var(--tw-bg-opacity));
  }
  .bg-rgb {
    --tw-bg-opacity: 1;
    background-color: rgba(23, 23, 24, var(--tw-bg-opacity));
  }
  .bg-rgba {
    --tw-bg-opacity: 0.5;
    background-color: rgba(23, 23, 25, var(--tw-bg-opacity));
  }
  .bg-hsl {
    --tw-bg-opacity: 1;
    background-color: rgba(0, 255, 0, var(--tw-bg-opacity));
  }
  .bg-hsla {
    --tw-bg-opacity: 0.3;
    background-color: rgba(128, 255, 128, var(--tw-bg-opacity));
  }
  .bg-hwb {
    --tw-bg-opacity: 1;
    background-color: rgba(102, 102, 102, var(--tw-bg-opacity));
  }
  .bg-hwba {
    --tw-bg-opacity: 0.7;
    background-color: rgba(102, 102, 102, var(--tw-bg-opacity));
  }
  .bg-srgb {
    --tw-bg-opacity: 1;
    background-color: rgba(178, 82, 82, var(--tw-bg-opacity));
  }
Tools / extend black / css / 0: |-
  .bg-black-300 {
    --tw-bg-opacity: 1;
    background-color: rgba(28, 28, 30, var(--tw-bg-opacity));
  }
  .bg-black {
    --tw-bg-opacity: 1;
    background-color: rgba(0, 0, 0, var(--tw-bg-opacity));
  }
  .bg-black-400 {
    --tw-bg-opacity: 1;
    background-color: rgba(34, 0, 0, var(--tw-bg-opacity));
  }
Tools / extend fontFamily with array / css / 0: |-
  .font-sans {
    font-family: Graphik,sans-serif;
  }
Tools / handle colors test / css / 0: |-
  .bg-discord {
    --tw-bg-opacity: 1;
    background-color: rgba(114, 137, 218, var(--tw-bg-opacity));
  }
  .bg-discord-100 {
    --tw-bg-opacity: 1;
    background-color: rgba(114, 137, 218, var(--tw-bg-opacity));
  }
  .bg-hex-7289da {
    --tw-bg-opacity: 1;
    background-color: rgba(114, 137, 218, var(--tw-bg-opacity));
  }
  .ring-offset-hex-1c1c1e {
    --tw-ring-offset-opacity: 1;
    --tw-ring-offset-color: rgba(28, 28, 30, var(--tw-ring-offset-opacity));
  }
  .ring-offset-gray-200 {
    --tw-ring-offset-opacity: 1;
    --tw-ring-offset-color: rgba(229, 231, 235, var(--tw-ring-offset-opacity));
  }
  .text-primary-color-red {
    --tw-text-opacity: 1;
    color: rgba(255, 0, 0, var(--tw-text-opacity));
  }
Tools / handle nest colors / colors / 0: |-
  .bg-my-custom-red {
    --tw-bg-opacity: 1;
    background-color: rgba(255, 0, 0, var(--tw-bg-opacity));
  }
  .bg-my {
    --tw-bg-opacity: 1;
    background-color: rgba(0, 0, 0, var(--tw-bg-opacity));
  }
  .bg-my-custom {
    --tw-bg-opacity: 1;
    background-color: rgba(255, 255, 255, var(--tw-bg-opacity));
  }
  .text-my-custom {
    --tw-text-opacity: 1;
    color: rgba(255, 255, 255, var(--tw-text-opacity));
  }
Tools / important string test / css / 0: |-
  @media (min-width: 640px) {
    #app .sm\:bg-black {
      --tw-bg-opacity: 1;
      background-color: rgba(0, 0, 0, var(--tw-bg-opacity));
    }
  }
Tools / important string test / css / 1: |-
  .dark #app .dark\:bg-white {
    --tw-bg-opacity: 1;
    background-color: rgba(255, 255, 255, var(--tw-bg-opacity));
  }
Tools / resolveConfig should extend correctly / order / 0: |-
  {
    "1": "1",
    "2": "2",
    "3": "3",
    "4": "4",
    "5": "5",
    "6": "6",
    "7": "7",
    "8": "8",
    "9": "9",
    "10": "10",
    "11": "11",
    "12": "12",
    "first": "-9999",
    "last": "9999",
    "none": "0",
    "lg": "44"
  }
Tools / resolveConfig should not overwrite theme function / extend-colors-red / 1: |-
  {
    "50": "#fef2f2",
    "100": "#fee2e2",
    "200": "#fecaca",
    "300": "#fca5a5",
    "400": "#f87171",
    "500": "#ef4444",
    "600": "#dc2626",
    "700": "#b91c1c",
    "800": "#991b1b",
    "900": "#7f1d1d",
    "custom": "#ea1851"
  }
Tools / resolveConfig should not overwrite theme function / extend-width / 0: |-
  {
    "0": "0px",
    "1": "0.25rem",
    "2": "0.5rem",
    "3": "0.75rem",
    "4": "1rem",
    "5": "1.25rem",
    "6": "1.5rem",
    "7": "1.75rem",
    "8": "2rem",
    "9": "2.25rem",
    "10": "2.5rem",
    "11": "2.75rem",
    "12": "3rem",
    "14": "3.5rem",
    "16": "4rem",
    "20": "5rem",
    "24": "6rem",
    "28": "7rem",
    "32": "8rem",
    "36": "9rem",
    "40": "10rem",
    "44": "11rem",
    "48": "12rem",
    "52": "13rem",
    "56": "14rem",
    "60": "15rem",
    "64": "16rem",
    "72": "18rem",
    "80": "20rem",
    "96": "24rem",
    "px": "1px",
    "0.5": "0.125rem",
    "1.5": "0.375rem",
    "2.5": "0.625rem",
    "3.5": "0.875rem",
    "xs": "20rem",
    "sm": "24rem",
    "md": "28rem",
    "lg": "32rem",
    "xl": "36rem",
    "2xl": "42rem",
    "3xl": "48rem",
    "4xl": "56rem",
    "5xl": "64rem",
    "6xl": "72rem",
    "7xl": "80rem",
    "1/2": "50%",
    "1/3": "33.333333%",
    "2/3": "66.666667%",
    "1/4": "25%",
    "2/4": "50%",
    "3/4": "75%",
    "1/5": "20%",
    "2/5": "40%",
    "3/5": "60%",
    "4/5": "80%",
    "1/6": "16.666667%",
    "2/6": "33.333333%",
    "3/6": "50%",
    "4/6": "66.666667%",
    "5/6": "83.333333%",
    "1/12": "8.333333%",
    "2/12": "16.666667%",
    "3/12": "25%",
    "4/12": "33.333333%",
    "5/12": "41.666667%",
    "6/12": "50%",
    "7/12": "58.333333%",
    "8/12": "66.666667%",
    "9/12": "75%",
    "10/12": "83.333333%",
    "11/12": "91.666667%",
    "auto": "auto",
    "full": "100%",
    "min": "min-content",
    "max": "max-content",
    "prose": "65ch",
    "screen": "100vw",
    "screen-sm": "640px",
    "screen-md": "768px",
    "screen-lg": "1024px",
    "screen-xl": "1280px",
    "screen-2xl": "1536px",
    "1/7": "14%",
    "2/7": "28%",
    "3/7": "42%",
    "4/7": "57%",
    "5/7": "71%",
    "6/7": "85%"
  }
Tools / shortcuts config object / shortcuts object / 0: |-
  .btn {
    color: white;
    border-radius: 0.5rem;
    font-weight: 600;
    padding-top: 0.5rem;
    padding-bottom: 0.5rem;
    padding-left: 1rem;
    padding-right: 1rem;
  }
  .btn:hover {
    --tw-bg-opacity: 1;
    background-color: rgba(4, 120, 87, var(--tw-bg-opacity));
    color: black;
  }
Tools / shortcuts config string / shortcuts string / 0: |-
  .btn {
    border-radius: 0.5rem;
    font-weight: 600;
    padding-top: 0.5rem;
    padding-bottom: 0.5rem;
    padding-left: 1rem;
    padding-right: 1rem;
    --tw-shadow: 0 4px 6px -1px rgb(0 0 0/0.1),0 2px 4px -2px rgb(0 0 0/0.1);
    --tw-shadow-colored: 0 4px 6px -1px var(--tw-shadow-color),0 2px 4px -2px var(--tw-shadow-color);
    -webkit-box-shadow: var(--tw-ring-offset-shadow,0 0 #0000),var(--tw-ring-shadow,0 0 #0000),var(--tw-shadow);
    box-shadow: var(--tw-ring-offset-shadow,0 0 #0000),var(--tw-ring-shadow,0 0 #0000),var(--tw-shadow);
  }
  .btn-green {
    --tw-bg-opacity: 1;
    background-color: rgba(16, 185, 129, var(--tw-bg-opacity));
    --tw-text-opacity: 1;
    color: rgba(255, 255, 255, var(--tw-text-opacity));
  }
  .btn-green:hover {
    --tw-bg-opacity: 1;
    background-color: rgba(4, 120, 87, var(--tw-bg-opacity));
  }
  @media (min-width: 1024px) {
    .lg\:btn {
      border-radius: 0.5rem;
      font-weight: 600;
      padding-top: 0.5rem;
      padding-bottom: 0.5rem;
      padding-left: 1rem;
      padding-right: 1rem;
      --tw-shadow: 0 4px 6px -1px rgb(0 0 0/0.1),0 2px 4px -2px rgb(0 0 0/0.1);
      --tw-shadow-colored: 0 4px 6px -1px var(--tw-shadow-color),0 2px 4px -2px var(--tw-shadow-color);
      -webkit-box-shadow: var(--tw-ring-offset-shadow,0 0 #0000),var(--tw-ring-shadow,0 0 #0000),var(--tw-shadow);
      box-shadow: var(--tw-ring-offset-shadow,0 0 #0000),var(--tw-ring-shadow,0 0 #0000),var(--tw-shadow);
    }
  }
'Tools / support color callback function #75 #256 / css / 0': |-
  .bg-transparent {
    background-color: transparent;
  }
  .bg-primary {
    --tw-bg-opacity: 1;
    background-color: hsla(var(--p, 259 94% 51%) / var(--tw-bg-opacity));
  }
  .bg-primary-focus {
    --tw-bg-opacity: 1;
    background-color: hsla(var(--pf, 259 94% 41%) / var(--tw-bg-opacity));
  }
  .text-primary-content {
    --tw-text-opacity: 1;
    color: hsla(var(--pc, 0 0% 100%) / var(--tw-text-opacity));
  }
Tools / transition property / css / 0: |-
  .btn {
    -webkit-transition-property: background-color;
    -o-transition-property: background-color;
    transition-property: background-color;
    -webkit-transition-timing-function: cubic-bezier(0.4, 0, 0.2, 1);
    -o-transition-timing-function: cubic-bezier(0.4, 0, 0.2, 1);
    transition-timing-function: cubic-bezier(0.4, 0, 0.2, 1);
    -webkit-transition-duration: 150ms;
    -o-transition-duration: 150ms;
    transition-duration: 150ms;
  }
Tools / variables theme config / css / 0: |-
  :root {
    --primary: #6366f1;
  }
